doc: clarify path.isAbsolute is not enough for mitigating path traversals (#57073)
nodejs/node: doc/api/path.md

* `path` {string}
* Returns: {boolean}

The `path.isAbsolute()` method determines if the literal (non-resolved) `path`

> is an absolute path. Therefore, it’s likely that you want to run it through
> `path.normalize()` before using it in `path.join()` or `path.isAbsolute()`.

Reviewer: ```suggestion
is an absolute path. Therefore, it’s likely that you want to run it through
`path.normalize()` if your plan is to use it as a subpath.
```

If the given `path` is a zero-length string, `false` will be returned.

For example, on POSIX:

```js
path.isAbsolute('/foo/bar');   // true
path.isAbsolute('/baz/..');    // true

> path.isAbsolute('/baz/../..'); // true (doesn’t resolve)

Reviewer 2: this path would still be absolute even if resolved, so i'm not sure what it adds to include it here? (you can't go higher than `/`, so `/baz/../..` would resolve to `/`, which is indeed absolute)

Author: `isAbsolute` is likely to be guarding traversals that could go higher, e.g. 
```js
  if (path.isAbsolute(myDir))
   path.join(ROOT, myDir)
```

Author: perhaps using another word for that comment such as `doesn’t simplify`. Not sure what’s the correct term for that

Reviewer 2: That seems a very confusing use case to me - an absolute path should never be prefixed by anything - but certainly if people are doing that then some kind of guidance seems beneficial.

Reviewer: In any case, even if there was a resolve, it would still be an absolute path
```suggestion
path.isAbsolute('/baz/../..'); // true
```

* `path` {string}
* Returns: {boolean}

The `path.isAbsolute()` method determines if the literal `path` is absolute.
Therefore, it’s not safe for mitigating path traversals without normalizing it.

```js
// Normalizing the subpath mitigates traversing above the mount directory
const subpath = '/foo/../..'
if (!path.isAbsolute(path.normalize(subpath))) {
  throw 'FORBIDDEN'
}

> const myPath = path.join(MOUNT_DIR, subpath)

Reviewer: This example is showing what NOT to do. E.g., if you take the following

```js
const MOUNT_DIR = "/home/example/";
const userProvidedPath = "/foo/../../../etc/passwd";

console.log(path.isAbsolute(path.normalize(userProvidedPath))); // true, i.e. it won't trigger the 'FORBIDDEN' error
console.log(path.join(MOUNT_DIR, userProvidedPath)); // /etc/passwd
```
Why not use the one I suggested? Explain to me what you don't like about it.

> The `path.isAbsolute()` method determines if the literal (non-resolved) `path` is an
> absolute path. Therefore, it’s likely that you want to run it through `path.normalize()`
> if your plan is to use it as a subpath in order to mitigate path traversals. For example:
> 
> ```js
> // Normalizing the subpath mitigates traversing above the mount directory.
> const mySubpath = path.normalize('/foo/../..');
> if (path.isAbsolute(mySubpath)) {
>   const myPath = path.join(MOUNT_DIR, mySubpath)
> }

Reviewer: I don't really understand why you would call `path.isAbsolute` at all. 
````suggestion
The `path.isAbsolute()` method determines if `path` is an absolute path. Being an
absolute path doesn't not guarentee it would be safe to treat as a subpath:

```js
const MOUNT_DIR = '/home/example/';
const userProvidedPath = '/foo/../../../etc/passwd';

console.log(path.isAbsolute(userProvidedPath)); // true
console.log(path.join(MOUNT_DIR, userProvidedPath)); // /etc/passwd

// Instead, call `path.normalize`:
console.log(path.join(MOUNT_DIR, path.normalize(userProvidedPath))); // /home/example/etc/passwd
````

Author: Flow control, e.g.
```js
const mySubpath = path.normalize('/foo/../..')
if (!path.isAbsolute(mySubpath))
  throw 'FORBIDDEN'
const myPath = path.join(MOUNT_DIR, mySubpath)
```

I agree it‘s not needed.

Author: Perhaps, instead of documenting, a better approach would be adding `path.isAbsoluteSubpath()`. That way, when autocompleting, users can tell there's something not obvious about `path.isAbsolute()`.

Something like,

```js

function isAbsoluteSubpath(path) {
  const dpath = decodeURIComponent(path)
  if (!dpath.startsWith('/'))
    return false
  const parts = dpath.split('/')
  let sum = 0
  for (const part of parts) {
    if (part === '')
      continue
    if (part === '..')
      sum--
    else
      sum++
    if (sum < 0)
      return false
  }
  return true
}

function is(path, expected) {
  console.assert(expected === isAbsoluteSubpath(path), path)
}

;[
  '.',
  './',
  '..',
  '/..',
  '../',
  '/home//../../..',
  '/home/user/../../..'
].forEach(p => {
  is(p, false)
  is(encodeURIComponent(p), false)
})

;[
  '/',
  '/home/user/docs',
  '/home/user/./docs',
  '/home/user//docs'
].forEach(p => {
  is(p, true)
  is(encodeURIComponent(p), true)
})
```

Reviewer: If the current behavior is confusing, let's update the doc to make them less confusing. Having clearer docs doesn't mean we can't add another API, that'd be a false dichotomy to think we can do one and not the other.

Would you like to take my suggestion so this PR can move forward?

```

If the given `path` is a zero-length string, `false` will be returned.

> On POSIX:

Reviewer: ```suggestion
For example, on POSIX:
```
